pilot/pkg/config/kube/ingress/conversion.go

```diff
@@ -105,22 +106,49 @@ func ConvertIngressV1alpha3(ingress v1beta1.Ingress, domainSuffix string) (model
 		Hosts: []string{"*"},
 	})
 
-	virtualService := &networking.VirtualService{
-		Hosts:    []string{"*"},
-		Gateways: []string{model.IstioIngressGatewayName},
+	gatewayConfig := model.Config{
+		ConfigMeta: model.ConfigMeta{
+			Type:      model.Gateway.Type,
+			Group:     model.Gateway.Group,
+			Version:   model.Gateway.Version,
+			Name:      ingress.Name + "-" + model.IstioIngressGatewayName,
+			Namespace: model.IstioIngressNamespace,
+			Domain:    domainSuffix,
+		},
+		Spec: gateway,
 	}
 
-	var httpRoutes []*networking.HTTPRoute
+	return gatewayConfig
+}
+
+func ConvertIngressVirtualService(ingress v1beta1.Ingress, domainSuffix string, ingressByHost map[string]*model.Config) {
+	// Ingress allows a single host - if missing '*' is assumed
+	// We need to merge all rules with a particular host across
+	// all ingresses, and return a separate VirtualService for each
+	// host.
+
 	for _, rule := range ingress.Spec.Rules {
 		if rule.HTTP == nil {
 			log.Infof("invalid ingress rule %s:%s for host %q, no paths defined", ingress.Namespace, ingress.Name, rule.Host)
 			continue
 		}
 
+		host := rule.Host
+		namePrefix := strings.Replace(host, ".", "-", -1)
+		if host == "" {
+			host = "*"
+		}
+		virtualService := &networking.VirtualService{
+			Hosts:    []string{},
+			Gateways: []string{model.IstioIngressGatewayName},
+		}
+
+		virtualService.Hosts = []string{host}
+
+		httpRoutes := []*networking.HTTPRoute{}
```

Comment:
it's more efficient and idiomatic to write this as var httpRoutes []*networking.HTTPRoute, as there's no need to create an empty array at this point and better to leave it as nil.